### PR TITLE
Change Android Permission Module to use promises

### PR DESCRIPTION
**Pull request by cmcewen**
This is a follow-up to #9292 @satya164

### CODE REVIEW

**Comment by facebook-github-bot-2** [comment]
By analyzing the blame information on this pull request, we identified @andreicoman11 and @spicyj to be potential reviewers.

**Comment by andreicoman11** [comment]
Mind rebasing this on top of https://github.com/facebook/react-native/pull/9292 once it lands? This lgtm, we can ship it afterwards

**Comment by facebook-github-bot-9** [comment]
@cmcewen updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@cmcewen updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@cmcewen updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@cmcewen updated the pull request.

File: Libraries/PermissionsAndroid/PermissionsAndroid.js [hunk 1]
```diff
@@ -119,36 +113,25 @@ class PermissionsAndroid {
    * and then shows the system permission dialog
    */
   requestPermission(permission: string, rationale?: Rationale) : Promise<boolean> {
-    return new Promise((resolve, reject) => {
```
Comment (satya164) [inline Libraries/PermissionsAndroid/PermissionsAndroid.js]: The code got a bit messy here, how about,

```js
async requestPermission(permission: string, rationale?: Rationale) : Promise<boolean> {
  if (rationale) {
    const shouldShowRationale = await Permissions.shouldShowRequestPermissionRationale(permission);

    if (shouldShowRationale) {
      return new Promise((resolve, reject) => {
        DialogManagerAndroid.showAlert(
            rationale,
            () => {
              DialogManagerAndroid.showAlert({
                message: 'Error Requesting Permissions'
              }, {}, {});
              reject();
            },
            resolve(Permissions.requestPermission(permission));
          );
      });
    }
  }
  return Permissions.requestPermission(permission);
}
```

I just wrote the code here, didn't test it.

File: Libraries/PermissionsAndroid/PermissionsAndroid.js [hunk 2]
```diff
@@ -119,36 +113,25 @@ class PermissionsAndroid {
...
 
       if (rationale) {
-        AndroidPermissions.shouldShowRequestPermissionRationale(
-          permission,
-          function(perm: string, shouldShowRationale: boolean) {
-            if (shouldShowRationale) {
-              DialogManagerAndroid.showAlert(
-                rationale,
-                () => {
-                  DialogManagerAndroid.showAlert({
```
Comment (satya164) [inline Libraries/PermissionsAndroid/PermissionsAndroid.js]: I think it's better left to the developer about what happens when requesting a permission fails.

**Comment by facebook-github-bot-9** [comment]
@cmcewen updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@cmcewen updated the pull request.

**Comment by facebook-github-bot-1** [comment]
@cmcewen updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9325/files/295bee5cb3240f816d175897881d67889ed2471b..7570815623d7997f6567dcd6a18f7dd937b18b22)

**Comment by facebook-github-bot-9** [comment]
@cmcewen updated the pull request.

**Comment by satya164** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-3** [comment]
Thanks for importing.If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3723568) to review internal test results.

**Pull request closed without merging**